### ISSUE TITLE
Add in cattle support for missing rancher-compose/docker-compose fields

### ISSUE THREAD

**Issue by deniseschannon** [rancher/rancher#4708]
As docker-compose/rancher-compose has added supported fields, we are missing the support within Rancher. 

After Rancher-compose has added support, there are additional areas that need to be updated

1. Adding the field to the container API
2. Adding the field to the exported docker-compose.yml of the stack
3. Adding the field to be read/imported by rancher-compose
4. UI (Once Cattle is complete, we can track the UI in a separate issue)

Known Fields that are missing.

Field | Description
---|---
 - [ ] --add-host=[]                   | Add a custom host-to-IP mapping (host:ip)
 - [ ]   --blkio-weight             |     Block IO (relative weight), between 10 and 1000
 - [ ]   --blkio-weight-device=[]      |  Block IO weight (relative device weight)
 - [ ]  --cpu-shares            |        CPU shares (relative weight)
 - [ ]  --cgroup-parent        |         Optional parent cgroup for the container
 - [ ]  --cpu-period             |       Limit CPU CFS (Completely Fair Scheduler) period
 - [ ]  --cpu-quota              |       Limit CPU CFS (Completely Fair Scheduler) quota
 - [ ]  --cpuset-cpus            |       CPUs in which to allow execution (0-3, 0,1)
 - [ ]  --cpuset-mems            |       MEMs in which to allow execution (0-3, 0,1)
 - [ ]  --device-read-bps=[]     |       Limit read rate (bytes per second) from a device
 - [ ]  --device-read-iops=[]    |       Limit read rate (IO per second) from a device
 - [ ]  --device-write-bps=[]    |       Limit write rate (bytes per second) to a device
 - [ ]  --device-write-iops=[]    |      Limit write rate (IO per second) to a device
 - [ ]  --dns-opt=[]            |        Set DNS options
 - [ ]  --dns-search=[]      |           Set custom DNS search domains
 - [ ]  --group-add=[]       |           Add additional groups to join
 - [ ]  --ip                     |       Container IPv4 address (e.g. 172.30.100.104)
 - [ ]  --ip6                 |          Container IPv6 address (e.g. 2001:db8::33)
 - [ ]  --ipc                 |          IPC namespace to use
 - [ ]  --isolation         |            Container isolation level
 - [ ]  --kernel-memory      |           Kernel memory limit
 - [ ]  --memory-reservation   |         Memory soft limit
 - [ ]  --memory-swap             |      Swap limit equal to memory plus swap: '-1' to enable unlimited swap
 - [ ]  --memory-swappiness=-1      |    Tune container memory swappiness (0 to 100)
 - [ ]  --net-alias=[]          |        Add network-scoped alias for the container
 - [ ]  --oom-kill-disable    |          Disable OOM Killer
 - [ ]  --oom-score-adj       |          Tune host's OOM preferences (-1000 to 1000)
 - [ ]  --read-only              |       Mount the container's root filesystem as read only
 - [ ]  --security-opt=[]      |         Security Options
 - [ ]  --shm-size             |         Size of /dev/shm, default value is 64MB
 - [ ]  --stop-signal=SIGTERM    |       Signal to stop a container, SIGTERM by default
- [ ]  --sysctl value        |        Sysctl options (default map[])
 - [ ]  --tmpfs=[]            |          Mount a tmpfs directory
 - [ ]  --ulimit=[]         |            Ulimit options
 - [ ]  --uts                 |          UTS namespace to usea

**Comment by mishak87** [comment]
Thanks for the list, at least I can add it to checks to avoid surprises. I just discovered that I'm missing ulimit myself.
IMO rancher-compose should throw at least warning on unsupported options.

**Comment by zhang8473** [comment]
Have you fixed this problem? Which rancher version shall I use?

**Comment by deniseschannon** [comment]
@zhang8473 Since the issue is still open, the fix is not in Rancher yet.

**Comment by giabao** [comment]
Because of this issue, when using [couchbase](https://hub.docker.com/r/couchbase/server/), I must manually run docker command from terminal.

How to manually run docker command so that rancher can recognize the service and show it in the UI? Ex, similar to this:
```
docker run -d \
--ulimit nofile=40960:40960 \
--ulimit core=100000000:100000000 \
--ulimit memlock=100000000:100000000 \
--net=host \
-v /mnt/sdb/couchbase/var:/opt/couchbase/var \
--label io.rancher.stack.name="couchbase" \
--label io.rancher.stack_service.name="couchbase/cb" \
--label io.rancher.project.name="couchbase" \
--label io.rancher.project_service.name="couchbase/cb" \
--label io.rancher.service.launch.config="io.rancher.service.primary.launch.config" \
--label io.rancher.service.deployment.unit="3bcf6f95-4db8-48ad-b3e9-d0972606ed4f" \
--label io.rancher.scheduler.affinity:host_label="cb=1" \
couchbase:4.5.0
```
(note the labels)

**Comment by deniseschannon** [comment]
@giabao You will not be able to launch a container and have it be a service just by adding those labels. Whenever you launch a container through docker CLI on the host, it will be a standalone container. Depending on what you want to do it as a service, you could use these steps to have it be part of a service as part of the DNS. 

Start a service in the UI. Add in a label `io.rancher.service.selector.container` with a value for a [selector link](http://docs.rancher.com/rancher/latest/en/labels/#selector-labels). 

Example: 
`io.rancher.service.selector.container: foo=bar`

Now, start up your container and add the following labels:

`io.rancher.container.network=true` - Allows the container to take advantage of the rancher DNS.
`foo=bar` - This label would make it so that the service would pick up the container as part of it's DNS.

**Comment by giabao** [comment]
@deniseschannon 

> ...the service would pick up the container as part of it's DNS

Could you explain more detail?

I have started couchbase as a standalone container:
`docker run -d --ulimit nofile=40960:40960` **--net=host --label foo=bar** `couchbase`

Then when I start a `cb-client` service from UI with label:
`io.rancher.service.selector.container: foo=bar`
How do I connect to couhbase from `cb-client`:
+ Do I need add label ` --label io.rancher.container.dns=true` when run couchbase container?
+ What is the domain name of couchbase as seen by cb-client? (how to `ping` to couchbase from inside cb-client shell?)

**Comment by deniseschannon** [comment]
@giabao In your case, you also start the couchbase container with `--label io.rancher.container.dns=true` and you'd want to actually use another service (`couchbase`) in the UI, that would have the selector label. Then, from `cb-client`, you would be able to ping `couchbase` and it should be able to reach the standalone container.

**Comment by deniseschannon** [comment]
The original task has been split into 4 issues:

Cattle Changes: #4708
Rancher-compose: #6280
Exporting yml: #6281
UI: #6282

**Comment by sangeethah** [comment]
API extends supports for adding ip,ip6,isolation and net-alias. 
But they are not currently used.

**Comment by StrongMonkey** [comment]
isolation is only for windows and other three seem to require to be attached to a network mode. (not fields for container) Had to verify it works before we can add them in agent.

**Comment by sangeethah** [comment]
Logged https://github.com/rancher/rancher/issues/6349 to track missing support for ip,ip6,isolation and net-alias.

**Comment by sangeethah** [comment]
Tested with rancher-version - v1.2.0-pre4-rc4
Validated that the following new fields can be set in launch_config when service is created and we are able to see these values being set in containers using docker inspect:
 "cpuShares":
 "blkioWeight":
"blkioDeviceOptions": 
 "cpuPeriod":
 "cpuQuota":
"cpuSet":
 "cpuSetMems": 
"dnsSearch": 
 "dnsOpt": 
 "groupAdd": 
"kernelMemory": 
"memoryReservation": 
 "memorySwap":
"memorySwappiness": 
"oomKillDisable": 
"oomScoreAdj":
"readOnly": 
"securityOpt":
 "shmSize":
"stopSignal":
"sysctls":
 "tmpfs": 
"ipcMode":
"uts":

**Comment by sangeethah** [comment]
Not able to validate "BlkioWeightDevice" option , since containers when deployed with this option, fails to start with error message  - ```
2016-10-21 00:26:07,576 ERROR [f008691a-07e0-419f-a0ba-248d09316f4d:4847] [instance:133->instanceHostMap:133] [instance.start->(InstanceStart)->instancehostmap.activate] [] [ServiceReplay-2] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [compute.instance.activate.reply;agent=1]: InstanceActivate: failed to activate instance: DoInstanceActivate: failed to start container: Error response from daemon: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"process_linux.go:291: setting cgroup config for ready process caused \\\"failed to write 1:3 20 to blkio.weight_device: write /sys/fs/cgroup/blkio/docker/4dff243769c0f219d1f92146e9aff02d60e15f24410369658d97e4311028e94a/blkio.weight_device: operation not supported\\\"\"\n" 
```
launch_config has the following value set for blkioDeviceOptions:

```  
  dev_opts = {
        '/dev/sda': {
            'readIops': 2000,
            'writeIops': 3000,
            'weight': 10
        },
```

**Comment by StrongMonkey** [comment]
Weight parameter depends on specific kernels. @sangeethah if the other four work fine, you probably close this one.

**Comment by sangeethah** [comment]
@StrongMonkey , Yes all other 4 parameters work and I see them using docker inspect command -
```
            "BlkioDeviceReadBps": [
                {
                    "Path": "/dev/null",
                    "Rate": 4000
                }
            ],
            "BlkioDeviceWriteBps": [
                {
                    "Path": "/dev/null",
                    "Rate": 200
                }
            ],
            "BlkioDeviceReadIOps": [
                {
                    "Path": "/dev/sda",
                    "Rate": 2000
                }
            ],
            "BlkioDeviceWriteIOps": [
                {
                    "Path": "/dev/sda",
                    "Rate": 3000
                }
            ],
```
Logged #6349 to track missing support for ip,ip6,isolation and net-alias.
Closing this issue.

**Comment by cabrinoob** [comment]
This issue is closed. Does it mean that the list of docker flags is included in cattle?

**Comment by courcelm** [comment]
I'm having this issue:   strconv.ParseInt: parsing "256m": invalid syntax

with this docker-compose file:

version: '2'
services:
  hub:
    image: selenium/hub:3.0.1
  chrome:
    image: selenium/node-chrome:3.0.1
    shm_size: 256m
    links:
    - hub:hub

**Comment by FabienHenon** [comment]
I'm using version 1.4.1 but I don't see `ulimit` in the UI. Is there something I didn't understand ?
